File: src/g1ipmlinkedlist/CS/IPMLinkedList.Algo/LinkedList.cs
// File: LinkedList.cs

namespace IPMLinkedList.Algo;

public class LinkedList
{
    public LinkedListNode? Head { get; set; }

    public void InsertNodeAtHead(LinkedListNode node)
    {
        if (Head == null)
        {
            Head = node;
        }
        else
        {
            node.Next = Head;
            Head = node;
        }
    }

    public void CreateLinkedList(int[] values)
    {
        for (int i = values.Length - 1; i >= 0; i--)
        {
            LinkedListNode newNode = new(values[i]);

            InsertNodeAtHead(newNode);
        }
    }

    public void DisplayLinkedList()
    {
        LinkedListNode? temp = Head;

        Write("[");
        while (temp != null)
        {
            Write(temp.Data);
            temp = temp.Next;
            if (temp != null)
            {
                Write(", ");
            }
        }
        WriteLine("]");
    }

    public static LinkedListNode? Reverse(LinkedListNode? head)
    {
        LinkedListNode? prev = null;
        LinkedListNode? current = head;

        while (current != null)
        {
            LinkedListNode? next = current.Next;
            current.Next = prev;
            prev = current;
            current = next;
        }

        return prev;
    }
}

File: src/g1ipmlinkedlist/Java/com/example/ipmlinkedlist/LinkedList.java
// File: LinkedList.java

package com.example.ipmlinkedlist;

public class LinkedList {
    LinkedListNode head;

    public void insertNodeAtHead(LinkedListNode node) {
        if (head == null) {
            head = node;
        } else {
            node.next = head;
            head = node;
        }
    }

    public void createLinkedList(int[] values) {
        for (int i = values.length - 1; i >= 0; i--) {
            LinkedListNode newNode = new LinkedListNode(values[i]);
            insertNodeAtHead(newNode);
        }
    }

    public void displayLinkedList() {
        LinkedListNode temp = head;
        System.out.print("[");
        while (temp != null) {
            System.out.print(temp.data);
            temp = temp.next;
            if (temp != null) {
                System.out.print(", ");
            }
        }
        System.out.println("]");
    }

    public static LinkedListNode reverse(LinkedListNode head) {
        LinkedListNode prev = null;
        LinkedListNode current = head;

        while (current != null) {
            LinkedListNode next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }

        return prev;
    }
}
